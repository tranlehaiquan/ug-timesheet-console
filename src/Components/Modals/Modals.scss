.base-modal__content {
  max-height: calc(100vh - 2rem);
  overflow-y: auto;
}
