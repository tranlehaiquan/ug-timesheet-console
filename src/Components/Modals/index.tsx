export { ConfirmationModal } from "./ConfirmationModal";
export type { ConfirmationModalProps } from "./ConfirmationModal";
export { BaseModal } from "./BaseModal";
export type { BaseModalProps } from "./BaseModal";
