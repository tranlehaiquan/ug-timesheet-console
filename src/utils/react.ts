import * as React from 'react'

type IdentityFunction = <T>(fn: T) => T

export const typedMemo: IdentityFunction = React.memo

export const getDisplayName = <T extends any = any>(WrappedComponent: React.ComponentType<T>): string => {
  return WrappedComponent.displayName || WrappedComponent.name || 'Component'
}
