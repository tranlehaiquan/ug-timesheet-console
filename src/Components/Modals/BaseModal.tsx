import * as React from 'react'
import { BaseModal as UIBaseModal, Icon, Button } from 'skedulo-ui'


import './Modals.scss'

export interface BaseModalProps {
  title: string
  isOpened: boolean
  onClose: () => void
  children: React.ReactNode
  className?: string
}

export const BaseModal: React.FC<BaseModalProps> = props => {
  let className = 'base-modal__content sk-p-6'
  if (props.className) {
    className = `${className} ${props.className}`
  }

  return props.isOpened ? (
    <UIBaseModal>
      <div className={ className }>
        <header className="sk-flex sk-justify-between sk-items-center sk-mb-8">
          <h3 style={ { fontSize: 20, color: '#223049' } }>{ props.title }</h3>
          <Button
            buttonType="transparent"
            onClick={ props.onClose }
            compact
            className="sk-p-1 sk-leading-none"
          >
            <Icon
              name="remove"
              size={ 12 }
              color="#485875"
              className="sk-cursor-pointer"
            />
          </Button>
        </header>
        { props.children }
      </div>
    </UIBaseModal>
  ) : null
}

export default BaseModal