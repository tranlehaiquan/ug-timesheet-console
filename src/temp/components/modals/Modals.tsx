import * as React from "react";
import { omit } from "lodash";

import { Portal } from "../portal/Portal";
import { Button } from "../buttons/button/Button";
import { ButtonGroup } from "../button-group/ButtonGroup";

export const BaseModal: React.SFC = ({ children, ...otherProps }) => {
  return (
    <Portal>
      {/* TODO: UPDATE FOLLOWING LINES ON DESIGN COMPONENT */}
      <div
        className="sk-fade-in sk-fixed sk-pin sked-modal-bg sk-flex sk-justify-center sk-items-center sk-font-sans"
        data-sk-name="sked-modal"
      >
        <div
          {...otherProps}
          className="sk-bg-white sk-text-sm sk-rounded-medium sk-absolute sked-modal-position-top sk-shadow-lg"
        >
          {children}
        </div>
      </div>
    </Portal>
  );
};

BaseModal.displayName = "BaseModal";

export interface IConfirmationModalProps {
  onConfirm: (event?: React.MouseEvent<HTMLButtonElement>) => void;
  onCancel: (event?: React.MouseEvent<HTMLButtonElement>) => void;
  confirmButtonText?: string;
  /**
   * Used to indicate the current onConfirm action is in progress.
   */
  useWorkingStateOnConfirm?: boolean;
}

export interface IConfirmationModalState {
  confirmActionLoading: boolean;
}

export class ConfirmationModal extends React.PureComponent<
  IConfirmationModalProps,
  IConfirmationModalState
> {
  constructor(props: IConfirmationModalProps) {
    super(props);
    this.state = {
      confirmActionLoading: false,
    };
  }

  handleConfirmButtonClick = () => {
    const { onConfirm, useWorkingStateOnConfirm } = this.props;

    /* Used to indicate loading state if the confirm button action takes time */
    if (useWorkingStateOnConfirm) {
      this.setState({ confirmActionLoading: true });
    }

    return onConfirm();
  };

  render() {
    const { children, onCancel, onConfirm, confirmButtonText, ...rest } =
      this.props;
    const otherProps = omit(rest, ["useWorkingStateOnConfirm"]);

    return (
      <BaseModal {...otherProps}>
        <div className="sk-pt-10 sk-px-6 sk-pb-6 sk-flex sk-flex-col sked-modal-small">
          <div className="sk-flex-grow">{children}</div>
          <div className="sk-text-right">
            <ButtonGroup>
              <Button
                data-sk-name="cancel-button"
                buttonType="transparent"
                onClick={onCancel}
              >
                Cancel
              </Button>
              <Button
                data-sk-name="confirm-button"
                buttonType="primary"
                onClick={this.handleConfirmButtonClick}
                loading={this.state.confirmActionLoading}
              >
                {confirmButtonText || "Yes"}
              </Button>
            </ButtonGroup>
          </div>
        </div>
      </BaseModal>
    );
  }
}
