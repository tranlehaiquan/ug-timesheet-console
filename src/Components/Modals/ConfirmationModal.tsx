import * as React from 'react'
import { Button, ButtonGroup } from 'skedulo-ui'
import { BaseModal } from './BaseModal'


export interface ConfirmationModalProps {
  title: string
  isOpened: boolean
  onClose: () => void
  onCancel?: () => void
  onConfirm: (message?: string) => void
  children: React.ReactNode
  confirmLabel: string
  className?: string
}

export const ConfirmationModal: React.FC<ConfirmationModalProps> = props => {
  const onCancel = () => {
    if (props.onCancel) {
      props.onCancel()
    }
    if (props.onClose) {
      props.onClose()
    }
  }

  const onConfirm = () => {
    if (props.onConfirm) {
      props.onConfirm()
    }
    if (props.onClose) {
      props.onClose()
    }
  }

  return props.isOpened ? (
    <BaseModal { ...props }>
      <div className="sk-flex-grow">{ props.children }</div>
      <div className="sk-text-right">
        <ButtonGroup>
          <Button
            data-sk-name="cancel-button"
            buttonType="transparent"
            onClick={ onCancel }
          >
            Cancel
          </Button>
          <Button
            data-sk-name="confirm-button"
            buttonType="primary"
            onClick={ onConfirm }
          >
            { props.confirmLabel || 'Confirm' }
          </Button>
        </ButtonGroup>
      </div>
    </BaseModal>
  ) : null
}
