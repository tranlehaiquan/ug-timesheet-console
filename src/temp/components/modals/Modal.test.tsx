import * as React from 'react'
import { noop } from 'lodash'
import { shallow, render } from 'enzyme'

import { BaseModal, ConfirmationModal } from './Modals'

describe('Modal', () => {
  describe('BaseModal', () => {
    test('renders', () => {
      const wrapper = shallow(<BaseModal>Some content</BaseModal>)
      expect(wrapper).toMatchSnapshot()
    })

    test('adds data-sk-name attibute', () => {
      const wrapper = shallow(<BaseModal>Some content</BaseModal>)
      expect(wrapper.find('[data-sk-name="sked-modal"]').length).toEqual(1)
    })
  })

  describe('ConfirmationModal', () => {
    test('renders', () => {
      const wrapper = shallow(<ConfirmationModal onCancel={ noop } onConfirm={ noop }>Some content</ConfirmationModal>)
      expect(wrapper).toMatchSnapshot()
    })

    test('adds data-sk-name attibutes', () => {
      const wrapper = shallow(<ConfirmationModal onCancel={ noop } onConfirm={ noop }>Some content</ConfirmationModal>)
      expect(wrapper.find('[data-sk-name="cancel-button"]').length).toEqual(1)
      expect(wrapper.find('[data-sk-name="confirm-button"]').length).toEqual(1)
    })

    test('calls confirm prop function', () => {
      const confirmSpy = jest.fn()
      const wrapper = shallow(<ConfirmationModal onCancel={ noop } onConfirm={ confirmSpy }>Some content</ConfirmationModal>)
      wrapper.find('[data-sk-name="confirm-button"]').simulate('click')
      expect(confirmSpy).toBeCalled()
    })

    test('calls cancel prop function', () => {
      const cancelSpy = jest.fn()
      const wrapper = shallow(<ConfirmationModal onCancel={ cancelSpy } onConfirm={ noop }>Some content</ConfirmationModal>)
      wrapper.find('[data-sk-name="cancel-button"]').simulate('click')
      expect(cancelSpy).toBeCalled()
    })

    test('calls confirm prop function with loading indicator', () => {
      const confirmSpy = jest.fn()
      const wrapper = shallow(<ConfirmationModal onCancel={ noop } onConfirm={ confirmSpy } useWorkingStateOnConfirm>Some content</ConfirmationModal>)
      wrapper.find('[data-sk-name="confirm-button"]').simulate('click')
      expect(confirmSpy).toBeCalled()
      expect(wrapper.state('confirmActionLoading')).toEqual(true)
    })

    test('uses supplied confirm button text', () => {
      const wrapper = shallow(<ConfirmationModal onCancel={ noop } onConfirm={ noop }  confirmButtonText="I work">Some content</ConfirmationModal>)
      expect(wrapper.find('[data-sk-name="confirm-button"]').dive().text()).toEqual('I work<LoadingSpinner />')
    })
  })
})
