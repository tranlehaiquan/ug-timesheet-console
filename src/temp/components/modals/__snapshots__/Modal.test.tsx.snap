// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal BaseModal renders 1`] = `
<Portal>
  <div
    className="sk-fade-in sk-fixed sk-inset-0  sked-modal-bg sk-flex sk-justify-center sk-font-sans"
    data-sk-name="sked-modal"
  >
    <div
      className="sk-slide-up sk-bg-white sk-text-sm sk-rounded-medium sk-absolute sked-modal-position-top sk-shadow-lg"
    >
      Some content
    </div>
  </div>
</Portal>
`;

exports[`Modal ConfirmationModal renders 1`] = `
<BaseModal>
  <div
    className="sk-pt-10 sk-px-6 sk-pb-6 sk-flex sk-flex-col sked-modal-small"
  >
    <div
      className="sk-flex-grow"
    >
      Some content
    </div>
    <div
      className="sk-text-right"
    >
      <ButtonGroup>
        <ForwardRef
          buttonType="transparent"
          data-sk-name="cancel-button"
          onClick={[Function]}
        >
          Cancel
        </ForwardRef>
        <ForwardRef
          buttonType="primary"
          data-sk-name="confirm-button"
          loading={false}
          onClick={[Function]}
        >
          Yes
        </ForwardRef>
      </ButtonGroup>
    </div>
  </div>
</BaseModal>
`;
